### PR TITLE
Add a new mode to benchmark the backend specifically.

### PR DESCRIPTION
**Pull request by Tilps**
Currently with a TODO to support multiple threads, but I really don't think I want to bother with that any time soon.

### CODE REVIEW

**Comment by Tilps** [comment]
NOTE: - bbench causes serious gpu load - use at own risk.

File: src/benchmark/bbench.cc [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                               "Maximum batch size to benchmark."};
+const OptionId kFenId{"fen", "", "Benchmark initial position FEN."};
+}  // namespace
+
+void BackendBenchmark::Run() {
+  OptionsParser options;
+  NetworkFactory::PopulateOptions(&options);
+  options.Add<IntOption>(kThreadsOptionId, 1, 128) = kDefaultThreads;
+
+  options.Add<IntOption>(kBatchesId, 1, 999999999) = 100;
```
Comment (Tilps) [inline src/benchmark/bbench.cc]: This default value might be a bit high.  For a big net 10 seems fine and runs much faster (obviously).

File: meson.build [hunk 2]
```diff
@@ -106,6 +106,7 @@ files += [
   'src/engine.cc',
   'src/version.cc',
   'src/benchmark/benchmark.cc',
+  'src/benchmark/bbench.cc',
```
Comment (mooskagh) [inline meson.build]: Not alphabetically sorted. :-P

File: src/main.cc [hunk 3]
```diff
@@ -59,6 +60,10 @@ int main(int argc, const char** argv) {
       // Benchmark mode.
       Benchmark benchmark;
       benchmark.Run();
+    } else if (CommandLine::ConsumeCommand("bbench")) {
```
Comment (mooskagh) [inline src/main.cc]: Maybe some more clear name? `backendbench` or `nnbench`.
Otherwise there will ask every day what `bbench` means.

Comment (Tilps) [reply src/main.cc]: Done.

File: meson.build (hunk 2 above)
Comment (Tilps) [reply meson.build]: Fixed.

File: src/main.cc [hunk 4]
```diff
@@ -59,6 +60,10 @@ int main(int argc, const char** argv) {
       // Benchmark mode.
       Benchmark benchmark;
       benchmark.Run();
+    } else if (CommandLine::ConsumeCommand("backendbench")) {
```
Comment (cn4750) [inline src/main.cc]: Don't forget to register the new mode so --help shows it as an option.

**Comment by Naphthalin** [comment]
Very useful feature!
For documentation: I ran the dx backend with my AMD RX Vega 56 yesterday, and it seems like at least for this card, multiples of 56 ist nearly always the best choice:

dx_backendbench_RX_Vega_56

**Comment by ASilver** [comment]
Here are a few benchmarks on a 2080 (normal) and 2080ti for 128, 256, 320 and 384 filters:

**128 filters**

_2080 (normal)_

Benchmark batch size 91 with inference average time 1.68911ms - throughput 53874.7 nps.
Benchmark batch size 92 with inference average time 1.67999ms - throughput 54762.1 nps.
Benchmark batch size 93 with inference average time 2.14624ms - throughput 43331.6 nps.

Benchmark batch size 183 with inference average time 2.7658ms - throughput 66165.4 nps.
Benchmark batch size 184 with inference average time 2.77098ms - throughput 66402.6 nps.
Benchmark batch size 185 with inference average time 3.24008ms - throughput 57097.4 nps.

Benchmark batch size 275 with inference average time 3.99533ms - throughput 68830.3 nps.
Benchmark batch size 276 with inference average time 4.00283ms - throughput 68951.3 nps.
Benchmark batch size 277 with inference average time 4.44985ms - throughput 62249.3 nps.

Benchmark batch size 367 with inference average time 5.21312ms - throughput 70399.3 nps.
Benchmark batch size 368 with inference average time 5.21963ms - throughput 70503.1 nps.
Benchmark batch size 369 with inference average time 5.63458ms - throughput 65488.4 nps.

Benchmark batch size 459 with inference average time 6.35907ms - throughput 72180.4 nps.
Benchmark batch size 460 with inference average time 6.38754ms - throughput 72015.2 nps.
Benchmark batch size 461 with inference average time 6.79572ms - throughput 67836.8 nps.

_2080ti_

Benchmark batch size 135 with inference average time 1.40734ms - throughput 95926 nps.
Benchmark batch size 136 with inference average time 1.42603ms - throughput 95369.8 nps.
Benchmark batch size 137 with inference average time 1.92129ms - throughput 71306.3 nps.

Benchmark batch size 271 with inference average time 2.46654ms - throughput 109870 nps.
Benchmark batch size 272 with inference average time 2.48117ms - throughput 109626 nps.
Benchmark batch size 273 with inference average time 2.87937ms - throughput 94812.3 nps.

Benchmark batch size 407 with inference average time 3.68205ms - throughput 110536 nps.
Benchmark batch size 408 with inference average time 3.64094ms - throughput 112059 nps.
Benchmark batch size 409 with inference average time 4.0841ms - throughput 100145 nps.

**256 filters**

_2080 (normal)_

Benchmark batch size 45 with inference average time 2.97138ms - throughput 15144.5 nps.
Benchmark batch size 46 with inference average time 2.97006ms - throughput 15487.9 nps.
Benchmark batch size 47 with inference average time 4.50368ms - throughput 10435.9 nps.

Benchmark batch size 91 with inference average time 5.18477ms - throughput 17551.4 nps.
Benchmark batch size 92 with inference average time 5.20807ms - throughput 17664.9 nps.
Benchmark batch size 93 with inference average time 6.77372ms - throughput 13729.5 nps.

Benchmark batch size 137 with inference average time 7.62701ms - throughput 17962.5 nps.
Benchmark batch size 138 with inference average time 7.65024ms - throughput 18038.7 nps.
Benchmark batch size 139 with inference average time 9.02551ms - throughput 15400.8 nps.

Benchmark batch size 183 with inference average time 9.9866ms - throughput 18324.6 nps.
Benchmark batch size 184 with inference average time 10.0017ms - throughput 18396.8 nps.
Benchmark batch size 185 with inference average time 11.4376ms - throughput 16174.7 nps.

Benchmark batch size 229 with inference average time 12.4206ms - throughput 18437.1 nps.
Benchmark batch size 230 with inference average time 12.4351ms - throughput 18496.1 nps.
Benchmark batch size 231 with inference average time 13.7487ms - throughput 16801.5 nps.

_2080ti_

Benchmark batch size 67 with inference average time 3.16604ms - throughput 21162.1 nps.
Benchmark batch size 68 with inference average time 3.16477ms - throughput 21486.5 nps.
Benchmark batch size 69 with inference average time 4.64244ms - throughput 14862.9 nps.

Benchmark batch size 135 with inference average time 5.82806ms - throughput 23163.8 nps.
Benchmark batch size 136 with inference average time 5.82149ms - throughput 23361.7 nps.
Benchmark batch size 137 with inference average time 7.46368ms - throughput 18355.5 nps.

Benchmark batch size 203 with inference average time 8.6718ms - throughput 23409.2 nps.
Benchmark batch size 204 with inference average time 8.6962ms - throughput 23458.5 nps.
Benchmark batch size 205 with inference average time 10.1585ms - throughput 20180.2 nps.

**320 filters**

_2080 (normal)_

Benchmark batch size 29 with inference average time 4.31965ms - throughput 6713.51 nps.
Benchmark batch size 30 with inference average time 4.32684ms - throughput 6933.46 nps.
Benchmark batch size 31 with inference average time 6.64218ms - throughput 4667.14 nps.

Benchmark batch size 71 with inference average time 8.58523ms - throughput 8270.02 nps.
Benchmark batch size 72 with inference average time 8.62972ms - throughput 8343.26 nps.
Benchmark batch size 73 with inference average time 11.1142ms - throughput 6568.19 nps.

Benchmark batch size 95 with inference average time 12.4515ms - throughput 7629.62 nps.
Benchmark batch size 96 with inference average time 12.4646ms - throughput 7701.8 nps.
Benchmark batch size 97 with inference average time 13.8577ms - throughput 6999.7 nps.

Benchmark batch size 121 with inference average time 14.9217ms - throughput 8109.02 nps.
Benchmark batch size 122 with inference average time 14.9585ms - throughput 8155.9 nps.
Benchmark batch size 123 with inference average time 17.0757ms - throughput 7203.23 nps.

Benchmark batch size 151 with inference average time 18.5106ms - throughput 8157.5 nps.
Benchmark batch size 152 with inference average time 18.5205ms - throughput 8207.11 nps.
Benchmark batch size 153 with inference average time 20.5324ms - throughput 7451.64 nps.

Benchmark batch size 183 with inference average time 22.1257ms - throughput 8270.91 nps.
Benchmark batch size 184 with inference average time 22.186ms - throughput 8293.51 nps.
Benchmark batch size 185 with inference average time 24.438ms - throughput 7570.18 nps.

Benchmark batch size 213 with inference average time 25.8529ms - throughput 8238.91 nps.
Benchmark batch size 214 with inference average time 25.893ms - throughput 8264.77 nps.
Benchmark batch size 215 with inference average time 27.9291ms - throughput 7698.06 nps.

Benchmark batch size 243 with inference average time 29.4104ms - throughput 8262.38 nps.
Benchmark batch size 244 with inference average time 29.4447ms - throughput 8286.72 nps.
Benchmark batch size 245 with inference average time 31.3773ms - throughput 7808.2 nps

_2080ti_

Benchmark batch size 43 with inference average time 4.75191ms - throughput 9048.99 nps.
Benchmark batch size 44 with inference average time 4.79926ms - throughput 9168.07 nps.
Benchmark batch size 45 with inference average time 6.99244ms - throughput 6435.52 nps.

Benchmark batch size 51 with inference average time 5.12934ms - throughput 9942.8 nps.
Benchmark batch size 52 with inference average time 5.16722ms - throughput 10063.4 nps.
Benchmark batch size 53 with inference average time 7.64452ms - throughput 6933.07 nps.

Benchmark batch size 95 with inference average time 9.54447ms - throughput 9953.4 nps.
Benchmark batch size 96 with inference average time 9.5536ms - throughput 10048.6 nps.
Benchmark batch size 97 with inference average time 11.5597ms - throughput 8391.2 nps.

Benchmark batch size 135 with inference average time 13.1462ms - throughput 10269.1 nps.
Benchmark batch size 136 with inference average time 13.1458ms - throughput 10345.5 nps.
Benchmark batch size 137 with inference average time 15.5778ms - throughput 8794.56 nps.

Benchmark batch size 179 with inference average time 17.5089ms - throughput 10223.4 nps.
Benchmark batch size 180 with inference average time 17.5194ms - throughput 10274.3 nps.
Benchmark batch size 181 with inference average time 19.6531ms - throughput 9209.73 nps.

Benchmark batch size 225 with inference average time 21.8174ms - throughput 10312.9 nps.
Benchmark batch size 226 with inference average time 21.8378ms - throughput 10349 nps.
Benchmark batch size 227 with inference average time 23.8501ms - throughput 9517.78 nps.

**384 filters**

_2080 (normal)_

Benchmark batch size 29 with inference average time 5.85924ms - throughput 4949.44 nps.
Benchmark batch size 30 with inference average time 5.86841ms - throughput 5112.12 nps.
Benchmark batch size 31 with inference average time 8.19766ms - throughput 3781.57 nps.

Benchmark batch size 59 with inference average time 8.78644ms - throughput 6714.89 nps.
Benchmark batch size 60 with inference average time 8.81344ms - throughput 6807.79 nps.
Benchmark batch size 61 with inference average time 13.8248ms - throughput 4412.37 nps.

Benchmark batch size 121 with inference average time 16.0038ms - throughput 7560.72 nps.
Benchmark batch size 122 with inference average time 16.0328ms - throughput 7609.38 nps.
Benchmark batch size 123 with inference average time 18.1496ms - throughput 6777.03 nps.

Benchmark batch size 183 with inference average time 23.6322ms - throughput 7743.67 nps.
Benchmark batch size 184 with inference average time 23.6732ms - throughput 7772.5 nps.
Benchmark batch size 185 with inference average time 25.858ms - throughput 7154.46 nps.

Benchmark batch size 213 with inference average time 27.4433ms - throughput 7761.45 nps.
Benchmark batch size 214 with inference average time 27.4771ms - throughput 7788.29 nps.
Benchmark batch size 215 with inference average time 29.4917ms - throughput 7290.19 nps.

Benchmark batch size 243 with inference average time 31.0962ms - throughput 7814.47 nps.
Benchmark batch size 244 with inference average time 31.1425ms - throughput 7834.95 nps.
Benchmark batch size 245 with inference average time 33.1414ms - throughput 7392.58 nps.

_2080ti_

Benchmark batch size 43 with inference average time 5.7733ms - throughput 7448.08 nps.
Benchmark batch size 44 with inference average time 5.80699ms - throughput 7577.08 nps.
Benchmark batch size 45 with inference average time 7.93602ms - throughput 5670.35 nps.

Benchmark batch size 87 with inference average time 9.35797ms - throughput 9296.89 nps.
Benchmark batch size 88 with inference average time 9.37752ms - throughput 9384.14 nps.
Benchmark batch size 89 with inference average time 14.0607ms - throughput 6329.69 nps.

Benchmark batch size 135 with inference average time 14.8307ms - throughput 9102.77 nps.
Benchmark batch size 136 with inference average time 14.8555ms - throughput 9154.88 nps.
Benchmark batch size 137 with inference average time 16.9027ms - throughput 8105.21 nps.

Benchmark batch size 179 with inference average time 19.155ms - throughput 9344.81 nps.
Benchmark batch size 180 with inference average time 19.1727ms - throughput 9388.35 nps.
Benchmark batch size 181 with inference average time 21.3509ms - throughput 8477.39 nps.

Benchmark batch size 225 with inference average time 23.7145ms - throughput 9487.88 nps.
Benchmark batch size 226 with inference average time 23.7519ms - throughput 9515.02 nps.
Benchmark batch size 227 with inference average time 25.8323ms - throughput 8787.46 nps.

**Comment by Naphthalin** [comment]
To add for a 2080 Super:

10x128:
Benchmark batch size 95 with inference average time 1.03627ms - throughput 91674.9 nps.
Benchmark batch size 96 with inference average time 1.04365ms - throughput 91984.9 nps.
Benchmark batch size 97 with inference average time 1.48372ms - throughput 65376.4 nps.

Benchmark batch size 191 with inference average time 1.8929ms - throughput 100903 nps.
**Benchmark batch size 192 with inference average time 1.89576ms - throughput 101279 nps.**
Benchmark batch size 193 with inference average time 2.36258ms - throughput 81690.4 nps.

Benchmark batch size 256 with inference average time 2.71108ms - throughput 94427.3 nps.

24x320:
Benchmark batch size 31 with inference average time 3.79358ms - throughput 8171.71 nps.
Benchmark batch size 32 with inference average time 3.78805ms - throughput 8447.62 nps.
Benchmark batch size 33 with inference average time 6.17158ms - throughput 5347.09 nps.

Benchmark batch size 75 with inference average time 8.14745ms - throughput 9205.34 nps.
**Benchmark batch size 76 with inference average time 8.17693ms - throughput 9294.44 nps.**
Benchmark batch size 77 with inference average time 10.6504ms - throughput 7229.78 nps.

Benchmark batch size 128 with inference average time 14.323ms - throughput 8936.68 nps.
Benchmark batch size 192 with inference average time 21.3242ms - throughput 9003.86 nps.
Benchmark batch size 256 with inference average time 28.4253ms - throughput 9006.06 nps.

**Comment by Tilps** [comment]
I can't repro ASilvers 2080 TI results with 384 filters, his NPS is way too high. 
@ASilver can you confirm which network file you used?

**Comment by ASilver** [comment]
@Tilps My apologies. I created a bunch of basic builds with unorthodox structures to see if the number of blocks affected the results (other than NPS), or if the break points remained at the same places, and they did. Since the focus was on where the break points were, I did not bother to add this. Thus I tested also 256x32, 256x40, 384x20, 384x30, etc. I also tested with multiplexing and roundrobin, and what I found out was that while the breakpoints do not change at all if both cards are the same, if multiplexing is used for a mixed card setup the result does not correspond to one card or the other, you get a whole different set of values.

**Comment by Naphthalin** [comment]
Separate post for separate GPU:
Titan RTX with cudnn-fp16 backend

10x128:

Benchmark batch size 143 with inference average time 1.18546ms - throughput 120628 nps.
Benchmark batch size 144 with inference average time 1.18791ms - throughput 121221 nps.
Benchmark batch size 145 with inference average time 1.63519ms - throughput 88674.6 nps.

Benchmark batch size 256 with inference average time 2.07439ms - throughput 123410 nps.

20x256:

Benchmark batch size 71 with inference average time 3.263ms - throughput 21759.1 nps.
Benchmark batch size 72 with inference average time 3.27128ms - throughput 22009.7 nps.
Benchmark batch size 73 with inference average time 4.8746ms - throughput 14975.6 nps.

Benchmark batch size 143 with inference average time 6.07074ms - throughput 23555.6 nps.
**Benchmark batch size 144 with inference average time 6.08634ms - throughput 23659.5 nps.**
Benchmark batch size 145 with inference average time 7.75278ms - throughput 18703 nps.

Benchmark batch size 216 with inference average time 9.1383ms - throughput 23636.8 nps.
Benchmark batch size 256 with inference average time 11.431ms - throughput 22395.3 nps.

24x320:

Benchmark batch size 55 with inference average time 4.40434ms - throughput 12487.7 nps.
**Benchmark batch size 56 with inference average time 4.43454ms - throughput 12628.1 nps.**
Benchmark batch size 57 with inference average time 6.82895ms - throughput 8346.82 nps.

Benchmark batch size 144 with inference average time 11.991ms - throughput 12009 nps.
Benchmark batch size 256 with inference average time 23.0515ms - throughput 11105.6 nps.

30x384:

Benchmark batch size 95 with inference average time 11.8937ms - throughput 7987.44 nps.
**Benchmark batch size 96 with inference average time 11.9213ms - throughput 8052.79 nps.**
Benchmark batch size 97 with inference average time 16.0412ms - throughput 6046.91 nps.

Benchmark batch size 144 with inference average time 19.0251ms - throughput 7568.95 nps.
Benchmark batch size 192 with inference average time 25.3714ms - throughput 7567.59 nps.
Benchmark batch size 240 with inference average time 31.5112ms - throughput 7616.34 nps.
Benchmark batch size 256 with inference average time 35.6141ms - throughput 7188.16 nps.

**Comment by cn4750** [comment]
Hardware: Nvidia RTX 2070 Super
| 128x10 | Batch Size | NPS |
| --- | --- | --- |
|| 80 | 67803 |
|| 160 | 80665 |
|| 240 | 84109 |
| 256x20 | Batch Size | NPS |
| --- | --- | --- |
|| 40 | 13704 |
|| 80 | 15542 |
|| 120 | 15734 |
|| 160 | 16155 |

**Comment by lp200** [comment]
Please add --min-batch-size option:grinning:

V100
384x30
Benchmark batch size 104 with inference average time 12.656ms - throughput 8217.43 nps.
Benchmark batch size 105 with inference average time 22.3796ms - throughput 4691.77 nps.
Benchmark batch size 212 with inference average time 24.9462ms - throughput 8498.29 nps.
Benchmark batch size 256 with inference average time 35.8606ms - throughput 7138.75 nps.
Benchmark batch size 320 with inference average time 37.9291ms - throughput 8436.8 nps.


320x24
Benchmark batch size 104 with inference average time 8.43903ms - throughput 12323.7 nps.
Benchmark batch size 212 with inference average time 16.702ms - throughput 12693.1 nps.
Benchmark batch size 320 with inference average time 24.9552ms - throughput 12823 nps.


256x20
Benchmark batch size 160 with inference average time 6.42852ms - throughput 24889.1 nps.
Benchmark batch size 256 with inference average time 12.0225ms - throughput 21293.4 nps.

**Comment by ankan-ban** [comment]
Here are results comparing dx12 and cudnn backends (both fp16) on Titan RTX:
chart

**Comment by ankan-ban** [comment]
Results for RTX 2060:
rtx_2060
...and with 128x10 network included:
rtx_2060_with_128x10


dx backend is currently very inefficient for smaller network sizes.

**Pull request merged**